### PR TITLE
 feat: support signing extrinsic of single source extension

### PR DESCRIPTION
**Pull request by AlexWZ0129**

### CODE REVIEW

File: src/commands/ext/sign.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import P2PSession from '../../util/p2pSession';
+
+export default class ExtSignCommand extends BaseWalletCommand {
+  static strict = false;
+
+  static description = `sign an extrinsic from single source extension
+`;
+
+  static args = [
+    {name: 'extrinsicString', required: true}
```
Comment (ianhe8x) [inline src/commands/ext/sign.ts]: description are needed so user know how to have the connection string

File: src/commands/ext/sign.ts [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+      const peerId = extrinsicRequest.peerId;
+      const secretKey = extrinsicRequest.secretKey;
+      const sessionId = extrinsicRequest.sessionId;
+
+      if (peerId && secretKey) {
+        const p2p = await P2PSession.connect(peerId, secretKey);
+
+        const signPayload: SignPayload = await new Promise(resolve => {
+          p2p.data$.subscribe(data => {
```
Comment (ianhe8x) [inline src/commands/ext/sign.ts]: i think `p2p.data$.pipe(first()).toPromise()` is better for readability

File: src/commands/ext/sign.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+          message: 'Do you want to sign this extrinsic?'
+        });
+
+        if (response.isConfirm) {
+          const wallet = await this.loadWallet(flags);
+          api.setSigner(wallet);
+
+          const hexSignature = await this.sign(signPayload, wallet, ext);
+
+          const peer = P2PSession.getPeer(peerId);
```
Comment (ianhe8x) [inline src/commands/ext/sign.ts]: i think you already have the peer connection object( `p2p`)

File: src/commands/ext/sign.ts (hunk 1 above)
Comment (AlexWZ0129) [reply src/commands/ext/sign.ts]: Okay

File: src/commands/ext/sign.ts (hunk 3 above)
Comment (AlexWZ0129) [reply src/commands/ext/sign.ts]: ah, right

File: src/commands/ext/sign.ts (hunk 2 above)
Comment (AlexWZ0129) [reply src/commands/ext/sign.ts]: yep

File: src/commands/ext/sign.ts (hunk 3 above)
Comment (AlexWZ0129) [reply src/commands/ext/sign.ts]: it actually gets p2p from P2PSession here. Let me just use p2p

File: src/commands/ext/sign.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    if (secretKey === undefined || secretKey === null) { this.consoleErrorAndExit('missing connectString'); }
+
+    const p2p = await P2PSession.connect(peerId, secretKey);
+
+    const data = await p2p.data$.pipe(first()).toPromise();
+    if (!this.isSignPayload(data)) {
+      p2p.destroy();
+      this.consoleErrorAndExit('unavailable SignPayload');
+    }
+    const signPayload: SignPayload = data as SignPayload;
```
Comment (ianhe8x) [inline src/commands/ext/sign.ts]: `: SignPayload` is not necessary, tsc can infer the type since you use `as SignPayload`.

**Pull request merged**